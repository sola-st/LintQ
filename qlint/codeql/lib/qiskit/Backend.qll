import python
import semmle.python.dataflow.new.DataFlow
import semmle.python.ApiGraphs
import qiskit.Register
import qiskit.Gate
import qiskit.Circuit

/** Simulator backend. */
class Backend extends DataFlow::CallCfgNode {
  // case: from qiskit import Aer; Aer.get_backend('qasm_simulator')
  Backend() {
    this = API::moduleImport("qiskit").getMember("Aer").getMember("get_backend").getACall() or
    this = API::moduleImport("qiskit").getMember("BasicAer").getMember("get_backend").getACall()
  }

  /** The name of the backend (according to the initialization string). */
  string getKind() {
    // case: from qiskit import Aer; Aer.get_backend('qasm_simulator')
    // > 'qasm_simulator'
    exists(StrConst str, DataFlow::LocalSourceNode strSource |
      strSource = this.getArg(0) and
      strSource.asExpr() = str
    |
      result = str.getText()
    )
  }

  /** If the simulator is statevector simulator. */
  predicate isStatevectorSimulator() {
    // case: from qiskit import Aer; Aer.get_backend('statevector_simulator')
    this.getKind().matches("%statevector%")
  }

  /** If the simulator is Unitary simulator. */
  predicate isUnitarySimulator() {
    // case: from qiskit import Aer; Aer.get_backend('unitary_simulator')
    this.getKind().matches("%unitary%")
  }

  /** Get a circuit run with this backend. */
  QuantumCircuit getACircuitToBeRun() {
    // case: Aer.get_backend('qasm_simulator').run(circuit)
    // > circuit
    exists(QuantumCircuit circuit, DataFlow::CallCfgNode runCall |
      runCall = this.getAnAttributeRead("run").getACall() and
      (
        circuit.flowsTo(runCall.getArg(0)) or
        circuit.flowsTo(runCall.getArgByName("circuits"))
      )
    |
      result = circuit
    )
    or
    // case: execute(qc, Aer.get_backend('qasm_simulator'), shots=1)
    // > qc
    exists(QuantumCircuit circuit, ExecuteCall executeCall |
      (
        circuit.flowsTo(executeCall.getArg(0)) or
        circuit.flowsTo(executeCall.getArgByName("experiments"))
      ) and
      (
        this.flowsTo(executeCall.getArg(1)) or
        this.flowsTo(executeCall.getArgByName("backend"))
      )
    |
      result = circuit
    )
  }

  /** Get a backend run with this backend. */
  BackendRun getABackendRun() {
    exists(BackendRun bkdRun |
      bkdRun.getBackend() = this
    |
      result = bkdRun
    )
  }

}

/** Backend run.
 *
 * This is a call to a backend run method.
 */
abstract class BackendRun extends DataFlow::CallCfgNode {

  abstract Backend getBackend();

}

/** A call to a backend run method. */
class BackendRunViaRunCall extends BackendRun {
  BackendRunViaRunCall() {
    exists(Backend backend, DataFlow::CallCfgNode runCall |
      runCall = backend.getAnAttributeRead("run").getACall()
    |
      this = runCall
    )
  }

  /** The backend that is run. */
  override Backend getBackend() {
    exists(Backend backend, DataFlow::CallCfgNode runCall |
      runCall = backend.getAnAttributeRead("run").getACall()
    |
      result = backend
    )
  }
}

/** A call to a backend execute method. */
class BackendRunViaExecuteCall extends BackendRun {
  BackendRunViaExecuteCall() {
    this instanceof ExecuteCall
  }

  /** The backend that is run. */
  override Backend getBackend() {
    exists(Backend backend |
      (
        backend.flowsTo(this.getArg(1)) or
        backend.flowsTo(this.getArgByName("backend"))
      )
    |
      result = backend
    )
  }
}

/** Result of a simulator run. */
class BackendResult extends DataFlow::CallCfgNode {
  BackendResult() {
    exists(BackendRun bkdRun |
      this = bkdRun.getAnAttributeRead("result").getACall()
      or
      this = bkdRun.getAnAttributeReference("result")
      or
      this = bkdRun.getAnAttributeWrite("result")
    )
  }

  /** The backend run that generated this result. */
  BackendRun getBackendRun() {
    exists(BackendRun bkdRun |
      // connect the run and its result
      this = bkdRun.getAnAttributeRead("result").getACall() or
      this = bkdRun.getAnAttributeReference("result") or
      this = bkdRun.getAnAttributeWrite("result")
    |
      result = bkdRun
    )
  }

}


/** Statevector returned by a result of a simulator run. */
class Statevector extends DataFlow::CallCfgNode {
  Statevector() {
    exists(BackendResult bkdResult |
      this = bkdResult.getAnAttributeRead("get_statevector").getACall()
    )
  }

  /** The backend run that generated this statevector. */
  BackendRun producedByBackendRun() {
    exists(BackendResult bkdResult, BackendRun bkdRun |
      // connect the run and its result
      bkdResult.getBackendRun() = bkdRun and
      // the result has been called to extract the statevector
      this = bkdResult.getAnAttributeRead("get_statevector").getACall()
    |
      result = bkdRun
    )
  }

}