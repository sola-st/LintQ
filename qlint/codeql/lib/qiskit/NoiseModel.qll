
import python
import semmle.python.dataflow.new.DataFlow
import semmle.python.ApiGraphs


class PauliError extends DataFlow::CallCfgNode {

    PauliError() {
        // OLDER API
        // from qiskit.providers.aer.noise.errors import pauli_error
        this = API::moduleImport("qiskit").getMember("providers")
            .getMember("aer").getMember("noise").getMember("errors")
            .getMember("pauli_error").getACall()
        or
        // NEWER API
        this = API::moduleImport("qiskit_aer").getMember("noise")
            .getMember("pauli_error").getACall()
    }

    StrConst getPauliString() {
        // pauli_error([('XY', p), ('I', 1 - p)])
        // extract 'XY' and 'I'
        exists(
            StrConst strConst
            |
            this.getArg(0).asCfgNode().getNode().contains(strConst)
            |
            result = strConst
            )
    }

    /** Check that the Pauli String are of different size.*/
    predicate arePauliStringSameSize() {
        // pauli_error([('XY', p), ('I', 1 - p)])
        // extract 'XY' and 'I'
        not exists(
            string str1, string str2
            |
            this.getPauliString().getText() = str1
            and
            this.getPauliString().getText() = str2
            |
            str1.length() != str2.length()
            )
    }


}
