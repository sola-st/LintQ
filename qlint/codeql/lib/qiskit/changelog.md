2023.03.07
- fix missing gates `crx` and `cry` in the `Gate` class
- improved the `isAppliedAfter` in the `Gate` class to remove false positives when the circuit is re-initialized during the path connecting the two gates.

2023.03.07
- added the getVar() function to the register class to get the variable name of the register.
- added the QubitUsed abstraction to keep track of each access to a qubit, considering also the register it belongs to.

2023.03.03
- refactoring model: java style method calls
- refactoring of register to share part of it.
- removed GateQuantumOperation, substituted with a predicate isMeasurement() on a general gate.

2023.02.28
- refactoring of the Gate class. Remove the tracking of the for loop.

2023.02.21
- improve the efficiency of get_a_target_qubit() of the gate class (remove extra QuantumCircuit variable)
- improved the precision of get_a_target_qubit() of the gate class to account also for qubits which are operated on in a loop (e.g. qc.h(i)).


2023.02.20
- added is_subcircuit() method to Circuit class (by checking if it flows to the compose() or append() method of another circuit object)
- improved get_total_num_qubits() and get_total_num_bits() methods of Circuit class (by considering also the registers added add_register() method)
- in Circuit refactored get_num_qubits_with_integers() and get_num_bits_with_integers() methods to return the static size of the circuit (i.e. QuantumCircuit(2,2) or n = 7; m = 2, QuantumCircuit(n, m)).
- improved get_num_bits() and get_num_qubits() methods of Register class (by considering also local flows of integer variables, e.g. x = 5; QuantumRegister(x))