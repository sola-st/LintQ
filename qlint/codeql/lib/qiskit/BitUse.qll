import qiskit.Circuit
import qiskit.Register
import semmle.python.dataflow.new.DataFlow
import semmle.python.dataflow.new.TaintTracking
import semmle.python.ApiGraphs

class EmptySetForString extends string {
  EmptySetForString() {
    this = "a" and this = "b" // this returns the empty set
  }
}

class EmptySetForInt extends int {
  EmptySetForInt() {
    this = 1 and this = 2 // this returns the empty set
  }
}

class EmptySetForRegisterV2 extends RegisterV2 {
  EmptySetForRegisterV2() {
    this.getName() != this.getName() // this returns the empty set
  }
}

class EmptySetForQuantumCircuit extends QuantumCircuit {
  EmptySetForQuantumCircuit() {
    this.getName() != this.getName() // this returns the empty set
  }
}

/** Usage of a qubit or bit. */
abstract class BitUse extends DataFlow::LocalSourceNode {
  int getAnIndex() {
    if exists(int i | i = this.getAnIndexIfAny())
    then exists(int i | i = this.getAnIndexIfAny() | result = i)
    else result = -1
  }

  string getARegisterName() {
    if exists(RegisterV2 reg | reg = this.getARegister())
    then exists(RegisterV2 reg | reg = this.getARegister() | result = reg.getName())
    else result = "anonymous register"
  }

  string getACircuitName() {
    if exists(QuantumCircuit circ | circ = this.getACircuit())
    then exists(QuantumCircuit circ | circ = this.getACircuit() | result = circ.getName())
    else result = "anonymous circuit"
  }

  abstract int getAnIndexIfAny();

  abstract RegisterV2 getARegister();

  abstract QuantumCircuit getACircuit();

  abstract string getAGateName();

  boolean equals(BitUse other) {
    if
      this.getARegisterName() = other.getARegisterName() and
      this.getAnIndexIfAny() = other.getAnIndexIfAny() and
      this.getACircuitName() = other.getACircuitName() and
      this.getARegister() = other.getARegister() and
      this.getACircuit() = other.getACircuit()
    then result = true
    else result = false
  }
}

/** Use of a qubit. */
abstract class QubitUse extends BitUse {
  override int getAnIndexIfAny() {
    // case: qc.h(0)
    // > 0
    // case: qc.h(quantum_register[7])
    // > 7
    if
      exists(IntegerLiteral lit, DataFlow::LocalSourceNode litSource |
        not this.asExpr() instanceof Subscript and
        litSource.asExpr() = lit and
        litSource.flowsTo(this)
      )
    then
      exists(IntegerLiteral lit, DataFlow::LocalSourceNode litSource |
        not this.asExpr() instanceof Subscript and
        litSource.asExpr() = lit and
        litSource.flowsTo(this)
      |
        result = lit.getValue()
      )
    else
      if exists(QuantumRegisterV2 qreg | qreg = this.getARegister())
      then
        // case: qc.h(quantum_register) with quantum_register = QuantumRegister(2)
        // > [0, 1]
        if exists(RegisterV2 reg | reg.flowsTo(this))
        then
          exists(RegisterV2 reg, int i |
            reg.flowsTo(this) and
            i in [0 .. reg.getSize() - 1]
          |
            result = i
          )
        else
          exists(
            QuantumRegisterV2 qreg, IntegerLiteral lit, DataFlow::LocalSourceNode litSource,
            DataFlow::LocalSourceNode indexDest
          |
            qreg = this.getARegister() and
            litSource.asExpr() = lit and
            indexDest.asExpr() = this.asExpr().(Subscript).getIndex() and
            litSource.flowsTo(indexDest)
          |
            result = lit.getValue()
          )
      else result instanceof EmptySetForInt
  }

  override RegisterV2 getARegister() {
    // case: qc.h(0)
    // > EmptySetForRegisterV2
    // case: qc.h(quantum_register[7])
    // > quantum_register
    // case: qc.h(quantum_register)
    // > quantum_register
    if
      exists(RegisterV2 reg |
        reg.getScope() = this.getScope() and
        reg.getName() = this.asExpr().(Subscript).getObject().toString()
      )
    then
      exists(RegisterV2 reg |
        reg.getScope() = this.getScope() and
        reg.getName() = this.asExpr().(Subscript).getObject().toString()
      |
        result = reg
      )
    else
      if exists(RegisterV2 reg | reg.flowsTo(this))
      then exists(RegisterV2 reg | reg.flowsTo(this) | result = reg)
      else result instanceof EmptySetForRegisterV2
  }
}

/** Use of a clbit. */
abstract class ClbitUse extends BitUse { }

/** Use of a qubit as attribute call on the circuit object. */
class QubitUseViaAttribute extends QubitUse {
  QubitUseViaAttribute() {
    exists(
      QuantumCircuit circ, GateSpecification gs, DataFlow::LocalSourceNode locSource,
      DataFlow::CallCfgNode call
    |
      // detect qc.h(0)
      call = circ.getAnAttributeRead(gs).getACall() and
      this = locSource
    |
      exists(int i | i = gs.getAnArgumentIndexOfQubit() |
        call.(API::CallNode).getParameter(i).getAValueReachingSink() = locSource
      )
      or
      exists(string kyw | kyw = gs.getAnArgumentNameOfQubit() |
        call.(API::CallNode).getKeywordParameter(kyw).getAValueReachingSink() = locSource
      )
    )
  }

  override string getAGateName() {
    exists(QuantumCircuit circ, GateSpecification gs, DataFlow::CallCfgNode call |
      // detect qc.h(0)
      call = circ.getAnAttributeRead(gs).getACall() and
      (
        exists(int i | i = gs.getAnArgumentIndexOfQubit() |
          call.(API::CallNode).getParameter(i).getAValueReachingSink() = this
        )
        or
        exists(string kyw | kyw = gs.getAnArgumentNameOfQubit() |
          call.(API::CallNode).getKeywordParameter(kyw).getAValueReachingSink() = this
        )
      )
    |
      result = gs
    )
  }

  override QuantumCircuit getACircuit() {
    exists(QuantumCircuit circ, GateSpecification gs, DataFlow::CallCfgNode call |
      // detect qc.h(0)
      call = circ.getAnAttributeRead(gs).getACall() and
      (
        exists(int i | i = gs.getAnArgumentIndexOfQubit() |
          call.(API::CallNode).getParameter(i).getAValueReachingSink() = this
        )
        or
        exists(string kyw | kyw = gs.getAnArgumentNameOfQubit() |
          call.(API::CallNode).getKeywordParameter(kyw).getAValueReachingSink() = this
        )
      )
    |
      result = circ
    )
  }
}

/** Use of a qubit as appended call on the circuit object. */
class QubitUseViaAppend extends QubitUse {
  QubitUseViaAppend() {
    exists(
      QuantumCircuit circ, GateSpecification gs, DataFlow::LocalSourceNode locSource,
      DataFlow::LocalSourceNode qubitListSource, DataFlow::CallCfgNode appendCall,
      DataFlow::CallCfgNode gateCall
    |
      // detect qc.append(CXGate(), [0, 1])
      appendCall = circ.getAnAttributeRead("append").getACall() and
      gateCall = appendCall.(API::CallNode).getParameter(0, "instruction").getAValueReachingSink() and
      gateCall =
        API::moduleImport("qiskit")
            .getMember("circuit")
            .getMember("library")
            .getMember(gs)
            .getACall() and
      qubitListSource = appendCall.(API::CallNode).getParameter(1, "qargs").getAValueReachingSink() and
      this = locSource
    |
      if qubitListSource.asExpr() instanceof List
      then qubitListSource.asExpr().(List).getAnElt() = locSource.asExpr()
      else qubitListSource.asExpr() = locSource.asExpr()
    )
  }

  override string getAGateName() {
    exists(
      QuantumCircuit circ, GateSpecification gs, DataFlow::LocalSourceNode qubitListSource,
      DataFlow::CallCfgNode appendCall, DataFlow::CallCfgNode gateCall
    |
      // detect qc.append(CXGate(), [0, 1])
      appendCall = circ.getAnAttributeRead("append").getACall() and
      gateCall = appendCall.(API::CallNode).getParameter(0, "instruction").getAValueReachingSink() and
      gateCall =
        API::moduleImport("qiskit")
            .getMember("circuit")
            .getMember("library")
            .getMember(gs)
            .getACall() and
      qubitListSource = appendCall.(API::CallNode).getParameter(1, "qargs").getAValueReachingSink() and
      // qubitListSource.asExpr().(List).getAnElt() = this.asExpr()
      if qubitListSource.asExpr() instanceof List
      then qubitListSource.asExpr().(List).getAnElt() = this.asExpr()
      else qubitListSource.asExpr() = this.asExpr()
    |
      result = gs
    )
  }

  override QuantumCircuit getACircuit() {
    exists(
      QuantumCircuit circ, GateSpecification gs, DataFlow::LocalSourceNode qubitListSource,
      DataFlow::CallCfgNode appendCall, DataFlow::CallCfgNode gateCall
    |
      // detect qc.append(CXGate(), [0, 1])
      appendCall = circ.getAnAttributeRead("append").getACall() and
      gateCall = appendCall.(API::CallNode).getParameter(0, "instruction").getAValueReachingSink() and
      gateCall =
        API::moduleImport("qiskit")
            .getMember("circuit")
            .getMember("library")
            .getMember(gs)
            .getACall() and
      qubitListSource = appendCall.(API::CallNode).getParameter(1, "qargs").getAValueReachingSink() and
      // qubitListSource.asExpr().(List).getAnElt() = this.asExpr()
      if qubitListSource.asExpr() instanceof List
      then qubitListSource.asExpr().(List).getAnElt() = this.asExpr()
      else qubitListSource.asExpr() = this.asExpr()
    |
      result = circ
    )
  }
}

class OldQubitUsedInteger extends IntegerLiteral {
  OldQubitUsedInteger() { exists(Gate gate | this = gate.getATargetQubit()) }

  Gate getGate() { exists(Gate gate | this = gate.getATargetQubit() | result = gate) }

  int getQubitIndex() { result = this.getValue() }

  QuantumRegister getQuantumRegister() {
    // OLD
    // exists(
    //     QuantumCircuit circ, QuantumRegister qr
    //     |
    //     circ = this.getGate().getQuantumCircuit() and
    //     qr = circ.getAQuantumRegister() and
    //     count(QuantumRegister reg | reg = circ.getAQuantumRegister() | reg) = 1
    //     |
    //     result = qr
    // )
    // or
    exists(QuantumRegister qr, SubscriptNode subscr |
      qr.getVar().getAUse() = subscr.getObject() and
      this = subscr.getNode().getIndex()
    |
      result = qr
    )
    // NEW
    // exists(
    //     QuantumRegister qr,
    //     QuantumCircuit circ,
    //     Gate gate
    //     |
    //     // this qubit is used in a gate
    //     this = gate.getATargetQubit()
    //     // the register belongs to the circuit
    //     and circ.getAQuantumRegister() = qr
    //     // the gate belongs to the circuit
    //     and circ = gate.getQuantumCircuit()
    //     and (
    //         // either it is the only register in the circuit
    //         // count(circ.getAQuantumRegister()) = 1
    //         // or
    //         (
    //             exists(
    //                 SubscriptNode subscr
    //                 |
    //                 // or we look for the register variable which
    //                 // is used to access the qubit (IntegerLiteral)
    //                 qr.getVar().getAUse() = subscr.getObject()
    //                 and this = subscr.getNode().getIndex()
    //             )
    //         )
    //     )
    //     |
    //     result = qr
    // )
  }
}

// GATE SPECIFICATIONS
abstract class GateSpecification extends string {
  GateSpecification() {
    this instanceof GateSpecificationObjectName or
    this instanceof GateSpecificationAttributeName
  }

  string getName() { result = this }

  /** The argument position pointing to a classical bit. */
  int getAnArgumentIndexOfClbit() {
    exists(int i |
      this.getNumberOfClbits() > 0 and
      i in [0 .. this.getNumberOfClbits() - 1]
    |
      // shift the index if there are qubits
      if this.getNumberOfQubits() > 0 then result = i + this.getNumberOfQubits() else result = i
    )
  }

  /** The named argument pointing to a classical bit. */
  string getAnArgumentNameOfClbit() { result instanceof EmptySetForString }

  /** The argument position pointing to a qubit. */
  int getAnArgumentIndexOfQubit() {
    exists(int i | this.getNumberOfQubits() > 0 and i in [0 .. this.getNumberOfQubits() - 1] |
      // shift the index if there are parameters
      if this.getNumberOfParams() > 0 then result = i + this.getNumberOfParams() else result = i
    )
  }

  /** The named argument pointing to a qubit. */
  string getAnArgumentNameOfQubit() { result instanceof EmptySetForString }

  /** The argument position pointing to a parameter. */
  int getAnArgumentIndexOfParam() {
    exists(int i |
      this.getNumberOfParams() > 0 and
      i in [0 .. this.getNumberOfParams() - 1]
    |
      result = i
    )
  }

  /** The named argument pointing to a parameter. */
  string getAnArgumentNameOfParam() { result instanceof EmptySetForString }

  /** Number of bits used in the gate. */
  int getNumberOfBits() { result = this.getNumberOfQubits() + this.getNumberOfClbits() }

  /** Number of parameters used in the gate. */
  int getNumberOfParams() { result = count(string s | s = this.getAnArgumentNameOfParam()) }

  /** Number of classical bits used in the gate. */
  int getNumberOfClbits() { result = count(string s | s = this.getAnArgumentNameOfClbit()) }

  /** Number of qubits used in the gate. */
  int getNumberOfQubits() { result = count(string s | s = this.getAnArgumentNameOfQubit()) }
}

class GateSpecificationAttributeName extends string {
  GateSpecificationAttributeName() {
    this in [
        // single bit operations
        "x", "y", "z", "h", "s", "sdg", "t", "tdg", "rx", "ry", "rz", "rv", "u1", "u2", "u3", "id",
        "i", "sx",
        // controlled operations
        "cx", "cnot", "cy", "cz", "ch", "cs", "csdg", "csx", "crz", "cry", "crx", "cu1", "cu3",
        "ccx", "ccz", "toffoli", "cswap", "fredkin", "mct", "rccx", "rcccx",
        // multi bit operations
        "rxx", "ryy", "rzz", "rzx", "swap", "iswap", "ms", "cr", "r", "rccx", "ecr",
        // measurements
        "measure", "measure_all"
      ]
  }
}

class GateSpecificationObjectName extends string {
  GateSpecificationObjectName() {
    this in [
        // single operations
        "XGate", "YGate", "ZGate", "HGate", "SGate", "SdgGate", "TGate", "TdgGate", "RXGate",
        "RYGate", "RZGate", "RVGate", "U1Gate", "U2Gate", "U3Gate", "IGate", "SXGate",
        // controlled operations
        "CXGate", "CYGate", "CZGate", "CHGate", "CSGate", "CSdgGate", "CSXGate", "CRZGate",
        "CRYGate", "CRXGate", "CU1Gate", "CU3Gate", "CCXGate", "CCZGate", "CSwapGate", "MCXGate",
        "RCCXGate", "RC3XGate",
        // multi bit operations
        "RXXGate", "RYYGate", "RZZGate", "RZXGate", "SwapGate", "iSwapGate", "MSGate", "CRGate",
        "RGate", "RCCXGate", "ECRGate",
      ]
  }
}

class GateSpecificationSingleQubitNoParam extends GateSpecification {
  GateSpecificationSingleQubitNoParam() {
    this in [
        "h", "x", "y", "z", "s", "sdg", "t", "tdg", "sx", "i", "id", "HGate", "XGate", "YGate",
        "ZGate", "SGate", "SdgGate", "TGate", "TdgGate", "SXGate", "IGate"
      ]
  }

  override string getAnArgumentNameOfQubit() { result = "qubit" }
}

class GateSpecificationRXGate extends GateSpecification {
  GateSpecificationRXGate() { this in ["rx", "RXGate"] }

  override string getAnArgumentNameOfQubit() { result = "qubit" }

  override string getAnArgumentNameOfParam() { result = "theta" }
}

class GateSpecificationRYGate extends GateSpecification {
  GateSpecificationRYGate() { this in ["ry", "RYGate"] }

  override string getAnArgumentNameOfQubit() { result = "qubit" }

  override string getAnArgumentNameOfParam() { result = "theta" }
}

class GateSpecificationRZGate extends GateSpecification {
  GateSpecificationRZGate() { this in ["rz", "RZGate"] }

  override string getAnArgumentNameOfQubit() { result = "qubit" }

  override string getAnArgumentNameOfParam() { result = "phi" }
}

class GateSpecificationRVGate extends GateSpecification {
  GateSpecificationRVGate() { this in ["rv", "RVGate"] }

  override string getAnArgumentNameOfQubit() { result = "qubit" }

  override string getAnArgumentNameOfParam() { result in ["vx", "vy", "vz"] }
}

// TODO CONTINUE GATES SINGLE QUBITS WITH PARAMS
class GateSpecificationCXGate extends GateSpecification {
  GateSpecificationCXGate() { this in ["cx", "CXGate", "cnot"] }

  override string getAnArgumentNameOfQubit() { result in ["control_qubit", "target_qubit"] }
}

class GateSpecificationCYGate extends GateSpecification {
  GateSpecificationCYGate() { this in ["cy", "CYGate"] }

  override string getAnArgumentNameOfQubit() { result in ["control_qubit", "target_qubit"] }
}

class GateSpecificationCZGate extends GateSpecification {
  GateSpecificationCZGate() { this in ["cz", "CZGate"] }

  override string getAnArgumentNameOfQubit() { result in ["control_qubit", "target_qubit"] }
}

class GateSpecificationCHGate extends GateSpecification {
  GateSpecificationCHGate() { this in ["ch", "CHGate"] }

  override string getAnArgumentNameOfQubit() { result in ["control_qubit", "target_qubit"] }
}
// TODO continue controlled gates and the rest
// QubitUsed
// derived from a Quantum gate used and its argument
// implicitely connected to a default QuantumRegister via an index
// e.g.  qc.h(0)
// or explicitely connected to a QuantumRegister via quantum_register[0]
// e.g. qc.h(quantum_register[0])
// .(Subscript).getIndex().(IntegerLiteral).getValue() = i
// class QubitUsed extends SubscriptNode {
//     // get a quantum register node which is used in a quantum gate
//     // qc.h(quantum_register[0])
//     QubitUsed() {
//         exists(
//             QuantumRegister qr, Variable reg_var
//             |
//             qr.getVar() = reg_var
//             and
//             reg_var.getAUse() = this.getObject()
//         ) or
//         exists(
//             QuantumRegister qr, Variable reg_var
//             |
//             qr.getVar() = reg_var
//             and
//             reg_var.getAUse() = this.getObject()
//         )
//     }
//     // get the index of the qubit used in the quantum gate
//     // qc.h(quantum_register[0])
//     int getQubitIndex() {
//         exists(
//             IntegerLiteral index
//             |
//             index = this.getNode().getIndex()
//             |
//             result = index.getValue()
//         )
//     }
//     QuantumRegister getQuantumRegister() {
//         exists(
//             QuantumRegister qr
//             |
//             qr.getVar().getAUse() = this.getObject()
//             |
//             result = qr
//         )
//     }
// }
// QubitSpace
// derived from a QuantumRegister declaration used in a QuantumCircuit
// e.g. quantum_register = QuantumRegister(2)
// Detector improvements
// find if there is any register with QubitSpace that is not used in any QubitUsed
