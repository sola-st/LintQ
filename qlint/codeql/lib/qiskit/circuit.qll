import python
import semmle.python.dataflow.new.DataFlow
import semmle.python.ApiGraphs
import qiskit.register


class QuantumCircuit extends DataFlow::CallCfgNode {
    QuantumCircuit() {
        this = API::moduleImport("qiskit").getMember("QuantumCircuit").getACall()
    }

    int get_num_qubits() {
        exists(IntegerLiteral num_qubits |
            num_qubits = this.getArg(0).asExpr() and
            result = num_qubits.getValue()
        ) or
        exists( QuantumRegister qntReg|
            // there is a dataflow between the quantum register and the quantum circuit
            exists(int i |
                qntReg.flowsTo(this.getArg(i)))
            and
            // the number of qubits is the number of qubits in the quantum register
            result = qntReg.get_num_qubits()
        )
    }

    int get_num_bits() {
        exists(IntegerLiteral num_bits |
            num_bits = this.getArg(1).asExpr() and
            result = num_bits.getValue()
        ) or
        exists( ClassicalRegister clsReg|
            // there is a dataflow between the classical register and the quantum circuit
            exists(int i |
                clsReg.flowsTo(this.getArg(i)))
            and
            // the number of bits is the number of bits in the classical register
            result = clsReg.get_num_bits()
        )
    }

    int get_total_num_bits() {
        result = sum(ClassicalRegister clsReg, int i |
            clsReg.flowsTo(this.getArg(i)) |
            clsReg.get_num_bits())
    }

    int get_total_num_qubits() {
        result = sum(QuantumRegister qntReg, int i |
            qntReg.flowsTo(this.getArg(i)) |
            qntReg.get_num_qubits())
    }

}