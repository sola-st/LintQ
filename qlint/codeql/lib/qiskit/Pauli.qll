import python
import semmle.python.dataflow.new.DataFlow
import semmle.python.ApiGraphs
import qiskit.NoiseModel


class PauliString extends StrConst {
    PauliString() {
        // it can be used in a PauliError
        exists(
            PauliError pauliError
            |
            pauliError.getPauliString() = this
        )
        or
        // it can be used in a Pauli object
        exists(
            DataFlow::CallCfgNode pauliObject
            |
            pauliObject = API::moduleImport("qiskit").getMember("circuit")
                        .getMember("QuantumCircuit").getMember("pauli").getACall()
            or
            // from qiskit.quantum_info import Pauli
            pauliObject = API::moduleImport("qiskit").getMember("quantum_info")
                        .getMember("Pauli").getACall()
            |
            this = pauliObject.getArg(0).asCfgNode().getNode()
        )
    }

    /** Check if the Pauli satisfy the regex ^[+-]?1?[ij]?[IXYZ]+$ */
    predicate isValid() {
        this.getText().regexpMatch("^[+-]?1?[ij]?[IXYZ]+$")
    }

}