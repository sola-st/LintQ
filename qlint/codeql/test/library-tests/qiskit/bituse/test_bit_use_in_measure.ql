import python
import qiskit.BitUse
import qiskit.Circuit

from QubitUse bu
where
  not bu.getLocation().getFile().getAbsolutePath().matches("%site-packages/qiskit/%") and
  bu.getAGate() instanceof Measurement
select bu, bu.getAGate(),
  "but use : circuit: " + bu.getACircuitName() + " - reg: " + bu.getARegisterName() + " - index: " +
    bu.getAnIndex() + " - gate: " + bu.getAGateName()
