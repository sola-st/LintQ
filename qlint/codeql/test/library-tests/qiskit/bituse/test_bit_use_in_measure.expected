| concept_gate_specification.py:51:12:51:18 | ControlFlowNode for Subscript | concept_gate_specification.py:51:1:51:28 | ControlFlowNode for Attribute() | but use : circuit: qc - reg: qreg - index: 0 - gate: measure |
| concept_gate_specification.py:52:1:52:16 | ControlFlowNode for Attribute() | concept_gate_specification.py:52:1:52:16 | ControlFlowNode for Attribute() | but use : circuit: qc - reg: anonymous register - index: 0 - gate: measure_all |
| concept_gate_specification.py:52:1:52:16 | ControlFlowNode for Attribute() | concept_gate_specification.py:52:1:52:16 | ControlFlowNode for Attribute() | but use : circuit: qc - reg: anonymous register - index: 1 - gate: measure_all |
| concept_gate_specification.py:52:1:52:16 | ControlFlowNode for Attribute() | concept_gate_specification.py:52:1:52:16 | ControlFlowNode for Attribute() | but use : circuit: qc - reg: anonymous register - index: 2 - gate: measure_all |
| concept_gate_specification.py:52:1:52:16 | ControlFlowNode for Attribute() | concept_gate_specification.py:52:1:52:16 | ControlFlowNode for Attribute() | but use : circuit: qc - reg: anonymous register - index: 3 - gate: measure_all |
| concept_gate_specification.py:52:1:52:16 | ControlFlowNode for Attribute() | concept_gate_specification.py:52:1:52:16 | ControlFlowNode for Attribute() | but use : circuit: qc - reg: anonymous register - index: 4 - gate: measure_all |
| concept_gate_specification.py:52:1:52:16 | ControlFlowNode for Attribute() | concept_gate_specification.py:52:1:52:16 | ControlFlowNode for Attribute() | but use : circuit: qc - reg: anonymous register - index: 5 - gate: measure_all |
| concept_gate_specification.py:52:1:52:16 | ControlFlowNode for Attribute() | concept_gate_specification.py:52:1:52:16 | ControlFlowNode for Attribute() | but use : circuit: qc - reg: anonymous register - index: 6 - gate: measure_all |
| concept_gate_specification.py:52:1:52:16 | ControlFlowNode for Attribute() | concept_gate_specification.py:52:1:52:16 | ControlFlowNode for Attribute() | but use : circuit: qc - reg: anonymous register - index: 7 - gate: measure_all |
| concept_gate_specification.py:52:1:52:16 | ControlFlowNode for Attribute() | concept_gate_specification.py:52:1:52:16 | ControlFlowNode for Attribute() | but use : circuit: qc - reg: anonymous register - index: -1 - gate: measure_all |
| concept_gate_specification.py:91:29:91:29 | ControlFlowNode for IntegerLiteral | concept_gate_specification.py:91:11:91:19 | ControlFlowNode for Measure() | but use : circuit: qc - reg: anonymous register - index: 0 - gate: Measure |
| concept_measure.py:3:6:3:23 | ControlFlowNode for QuantumRegister() | concept_measure.py:21:1:21:18 | ControlFlowNode for Attribute() | but use : circuit: qc - reg: qr - index: 0 - gate: measure |
| concept_measure.py:3:6:3:23 | ControlFlowNode for QuantumRegister() | concept_measure.py:21:1:21:18 | ControlFlowNode for Attribute() | but use : circuit: qc - reg: qr - index: 1 - gate: measure |
| concept_measure.py:8:12:8:16 | ControlFlowNode for Subscript | concept_measure.py:8:1:8:24 | ControlFlowNode for Attribute() | but use : circuit: qc - reg: qr - index: 0 - gate: measure |
| concept_measure.py:11:12:11:16 | ControlFlowNode for Subscript | concept_measure.py:11:1:11:20 | ControlFlowNode for Attribute() | but use : circuit: qc - reg: qr - index: 0 - gate: measure |
| concept_measure.py:12:12:12:12 | ControlFlowNode for IntegerLiteral | concept_measure.py:12:1:12:20 | ControlFlowNode for Attribute() | but use : circuit: qc - reg: anonymous register - index: 0 - gate: measure |
| concept_measure.py:16:12:16:12 | ControlFlowNode for IntegerLiteral | concept_measure.py:16:1:16:16 | ControlFlowNode for Attribute() | but use : circuit: qc - reg: anonymous register - index: 0 - gate: measure |
